src/coreComponents/constitutive/fluid/MultiFluidUtils.hpp

```diff
@@ -55,47 +55,41 @@ template< typename T, int DIM, int USD, int USD_DC >
 struct MultiFluidVarSlice
 {
   internal::ArraySliceOrRef< T, DIM, USD > value;        /// variable value
-  internal::ArraySliceOrRef< T, DIM, USD > dPres;        /// derivative w.r.t. pressure
-  internal::ArraySliceOrRef< T, DIM, USD > dTemp;        /// derivative w.r.t. temperature
-  internal::ArraySliceOrRef< T, DIM + 1, USD_DC > dComp; /// derivative w.r.t. composition
+  internal::ArraySliceOrRef< T, DIM + 1, USD_DC > derivs; /// derivative w.r.t. pressure, temperature, compositions
```

Comment:
This is where (along with `View` class below) where we would potentially add `dP`, `dT` and `dC` methods as described in https://github.com/GEOSX/GEOSX/pull/1225#discussion_r797828122 and https://github.com/GEOSX/GEOSX/pull/1225#discussion_r807554325 if we agree on that idea. The raw slice `derivs` need to remain accessible though, because it must be passed directly to `applyChainRule`.

This would somewhat decrease the flexibility of these classes however. For example, currently they are usable both with and without temperature derivatives - the owner decides whether or not it is present by resizing and then using the `derivs` array accordingly. By adding a `dT()` method we are forcing the temperature derivative to be always present. These classes could also be used outside of `MultiFluid` classes, anywhere one need to bundle values with a bunch of derivatives - for example in `RelativePermeabilityBase` and `CapillaryPressureBase`, physics solvers, etc.

For that reason I'd propose a refactoring in which this family of classes is moved into some shared place (`common` probably?) under a different name (`ValuesAndDerivatives[View|Slice]`?), while here we create MultiFluid-specific wrappers that either inherit from or contain the above classes and add the specialized access methods. I would propose maybe for this to be a separate PR.

I am also not opposed to just keeping the use of `Deriv::dP`, etc. as explicit offsets for the foreseeable future.

Reply:
I perceive this pattern as really similar (in nature) to the `extrinsicMeshData` wrappings of https://github.com/GEOSX/GEOSX/pull/1750 (the current situation brings additional complexity).

There would be an extremely carefully designed implementation (in the current case, some `CompositionalVarView`, in the case of https://github.com/GEOSX/GEOSX/pull/1750, the `*StencilAccessors`), that may be "_specialized_" depending on the context. That way, if you do not want the `dT`, then don't ask your smart implementation for space to store `dT`, and don't provide the `dT` accessors. Therefore why would we lose flexibility?

For example, it's fine for the underlying implementation to provide variadic accessors `real64 & dT( INDICES ... indices )` but providing a more constrained `real64 & dT( int a, int b, int c )` member function will be simpler in the end.

Reply:
I would vote for the following plan:
1. We merge this PR so that I can then finish and merge #1755 using the derivative format implemented in this PR
2. Once #1755 is merged, we work on implementing the `ValuesAndDerivatives` getters for the constitutive models and solvers

Even though this plan implies a transition period during which `Deriv::dP` etc will be used, it will allow me to start testing #1225 soon, which is important.

Regarding the idea of implementing getters (`dP()`, `dT()`, `dC()`) to get the derivatives in the kernels, will it slow down the code a little bit? Or not significantly?

Reply:
> That way, if you do not want the dT, then don't ask your smart implementation for space to store dT, and don't provide the dT accessors. Therefore why would we lose flexibility?

What I meant is, if we just added `dT()` getter to this class, it would become unusable in context where there is no temperature dependence. But of course as you said, if we provide appropriately specialized wrappers for different contexts, there will be no issues :+1: 

> I would vote for the following plan:

Sounds good to me!

> Regarding the idea of implementing getters (`dP()`, `dT()`, `dC()`) to get the derivatives in the kernels, will it slow down the code a little bit? Or not significantly?

I expect them to be inlined 100% of the time, so no slowdown, if only a tiny compilation time hit.

Reply:
Let's go! 🚗 💨 

> I would vote for the following plan:

If I were you, I would open an issue with multiple tick boxes, references to meaningful comments, etc.